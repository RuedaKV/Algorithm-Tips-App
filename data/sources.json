{
    "federal": [
        "Federal Agency - Executive",
        "Federal Agency - Judicial",
        "Federal Agency - Legislative"
    ],
    "regional": [
        "State/Local Govt",
        "Interstate Agency",
        "Native Sovereign Nation",
        "Independent Intrastate Agency"
    ],
    "local": [
        "City",
        "County"
    ]
}
