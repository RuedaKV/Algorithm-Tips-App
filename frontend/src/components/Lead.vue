<template>
  <div class="lead-box">
    <b-card no-body :header-bg-variant="lead.flagged ? 'warning' : 'default'">
      <template v-slot:header>
        <div class="d-flex justify-content-between align-items-center">
          <router-link :to="page_url" v-if="headerLink" class="header-link">
            <h5>{{ lead.name }}</h5>
          </router-link>
          <h5 v-else>{{ lead.name }}</h5>
          <b-button size="sm" @click="setFlag" v-if="!lead.flagged" class="flag-button">
            <b-spinner small class="flag-pending" v-show="flagPending" />Add Flag
          </b-button>
          <b-button size="sm" @click="unsetFlag" v-else class="flag-button">
            <b-spinner small class="flag-pending" v-show="flagPending" />Remove Flag
          </b-button>
        </div>
      </template>
      <b-card-body>
        <p class="quote">{{ lead.description }}</p>
        <!-- prettyhtml-preserve-whitespace -->
        <div class="source">
          <external-link :href="lead.link">{{ lead.link }}</external-link>&nbsp;
          (<external-link class="cache-link" v-if="cache" :href="cache">Cache</external-link>)
        </div>
        <!-- prettyhtml-preserve-whitespace -->
        <div class="found-by">Found via a search for &ldquo;<external-link :href="query_url">{{ lead.query_term }}</external-link>&rdquo; on {{ discovered }}. Published here on {{ published }}.</div>

        <h5>Additional Info</h5>
        <dl class="row">
          <dt class="col-sm-3 text-right">Jurisdiction</dt>
          <dd class="col-sm-9">{{ lead.jurisdiction }}</dd>
          <dt class="col-sm-3 text-right">Agency</dt>
          <dd class="col-sm-9">{{ lead.source }}</dd>
          <dt class="col-sm-3 text-right">Main Topics</dt>
          <dd class="col-sm-9">{{ lead.topic }}</dd>
          <dt class="col-sm-3 text-right">People &amp; Organizations</dt>
          <dd class="col-sm-9">
            <template v-if="people_orgs.length > 0">{{ people_orgs.join(", ") }}</template>
            <template v-else>
              <em class="text-muted">None found</em>
            </template>
          </dd>
        </dl>

        <h5 v-if="ratings_split.length">Crowd Ratings</h5>
      </b-card-body>
      <b-list-group flush>
        <!-- TODO: these bits are slowing down renders. not sure how much effort is worth spending on the perf here -->
        <b-list-group-item v-for="rating in ratings_split" :key="rating.title">
          <div v-b-toggle:[id+rating.title]>
            <b-icon-chevron-right class="when-closed" />
            <b-icon-chevron-down class="when-open" />
            <span class="rating-title">{{ rating.title }}</span>
            <b-progress :max="4" class="w-25 float-right inline-bar" variant="info">
              <b-progress-bar :value="rating.score - 1"></b-progress-bar>
            </b-progress>
            <span class="score-display float-right">{{ rating.score.toFixed(1) }} / 5</span>
          </div>

          <b-collapse :id="id+rating.title">
            <b-table-lite
              v-if="rating.comments.length"
              class="comment-table"
              striped
              :items="rating.comments"
            ></b-table-lite>
            <span v-else>
              <em class="text-muted">No rating information available.</em>
            </span>
          </b-collapse>
        </b-list-group-item>
      </b-list-group>
    </b-card>
    <b-modal
      id="flag-remove-dialog"
      v-if="confirmRemove"
      ref="delete-modal"
      centered
      title="Are you sure?"
      ok-title="Remove Flag"
      ok-variant="danger"
      @hidden="cancelRemove"
      @ok="commitRemove"
      :visible="pendingDeletion"
    >
      <p>Removing this flag will immediately remove it from this page. Are you sure?</p>
    </b-modal>
  </div>
</template>

<script>
import moment from "moment";
import { mapGetters, mapActions } from "vuex";
import externalLink from "./external-link.vue";
import sentence_case from "../sentence-case";

export default {
  name: "Lead",
  components: {
    "external-link": externalLink
  },
  props: { id: Number, headerLink: Boolean, confirmRemove: Boolean },
  data: () => {
    return {
      flagPending: false,
      pendingDeletion: false
    };
  },
  methods: {
    ...mapActions({
      updateFlag: `leads/updateFlag`
    }),
    redirect_login() {
      this.$router.push({
        path: "/login",
        query: {
          redirect: this.$route.fullPath
        }
      });
    },
    async setFlag() {
      if (!this.signedIn) {
        this.redirect_login();
        return;
      }

      try {
        this.flagPending = true;
        await this.updateFlag({ id: this.id, flag: true });
      } catch (err) {
        console.error("Unable to set flag", err);
      } finally {
        this.flagPending = false;
      }
    },
    unsetFlag() {
      if (!this.signedIn) {
        this.redirect_login();
        return;
      }

      if (this.confirmRemove) {
        this.pendingDeletion = true;
      } else {
        this.commitRemove();
      }
    },
    cancelRemove() {
      this.pendingDeletion = false;
    },
    async commitRemove() {
      this.pendingDeletion = false;
      try {
        this.flagPending = true;
        await this.updateFlag({ id: this.id, flag: false });

        this.$emit("remove-flag", this.id);
      } catch (err) {
        console.error("Unable to unset flag", err);
      } finally {
        this.flagPending = false;
      }
    }
  },
  computed: {
    ...mapGetters({
      find: "leads/find",
      signedIn: "user/signedIn"
    }),
    lead() {
      return this.find(this.id);
    },
    page_url() {
      return `/lead/${this.id}`;
    },
    cache() {
      return `https://algorithm-tips.s3.us-east-2.amazonaws.com/documents/${this.id}.${this.lead.document_ext}`;
    },
    query_url() {
      return `https://www.google.com/search?hl=en&q=%22${encodeURIComponent(
        this.lead.query_term
      )}%22+site%3A.gov+-site%3A.nih.gov&as_qdr=w1&lr=en&num=100`;
    },
    people_orgs() {
      const people = JSON.parse(this.lead.people) || {};
      const orgs = JSON.parse(this.lead.organizations) || {};

      const merged = Object.entries(people).concat(Object.entries(orgs));
      merged.sort(([, a], [, b]) => b - a);

      return merged.map(([name]) => name);
    },
    discovered() {
      const dt = moment(this.lead.discovered_dt);

      return dt.format("MMMM Do, YYYY");
    },
    published() {
      const dt = moment(this.lead.published_dt);

      return dt.format("MMMM Do, YYYY");
    },
    ratings_split() {
      const KEYS = ["controversy", "magnitude", "societal_impact", "surprise"];
      const LABELS = [
        "Potential for Controversy",
        "Number of People Impacted",
        "Negative Societal Impact",
        "Surprising, Unusual, or Unexpected"
      ];
      const mapper = obj =>
        KEYS.map((key, ix) => {
          return {
            category: key,
            label: LABELS[ix],
            score: 5 - obj[key] + 1,
            comment: sentence_case(obj[`${key}_explanation`])
          };
        });

      const init = obj => {
        return {
          title: obj.label,
          total_score: 0,
          get score() {
            return this.total_score / this.comments.length;
          },
          comments: []
        };
      };

      return Object.values(
        this.lead.ratings.flatMap(mapper).reduce((result, obj) => {
          if (!result[obj.category]) {
            result[obj.category] = init(obj);
          }

          result[obj.category].total_score += obj.score;
          result[obj.category].comments.push({
            score: obj.score,
            comment: obj.comment
          });
          return result;
        }, {})
      );
    }
  }
};
</script>

<style lang="stylus">
.lead-box {
  margin-bottom: 2em;

  p.quote {
    margin-bottom: 0;
  }

  .card-header h5 {
    margin-bottom: 0;
    display: inline;
  }

  .card-body h5 {
    margin-top: 1rem;
  }

  .card-body {
    padding-bottom: 0;
  }

  dd {
    margin-bottom: 0.1rem;
  }

  .inline-bar {
    display: inline-flex;
    margin-top: 0.25rem;
  }

  .score-display {
    margin-right: 1em;
  }

  .comment-table {
    margin-top: 1em;
  }

  .found-by {
    font-style: italic;
  }

  .collapsed > .when-open, .not-collapsed > .when-closed {
    display: none;
  }

  .rating-title {
    padding-left: 1em;
  }

  a.header-link {
    color: initial;
  }

  .flag-pending {
    margin-right: 1em;
  }

  .flag-button {
    min-width: 7em;
  }
}
</style>