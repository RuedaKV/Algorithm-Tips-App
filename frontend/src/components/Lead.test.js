import { shallowMount, createLocalVue } from '@vue/test-utils';
import { BootstrapVue, IconsPlugin } from 'bootstrap-vue';
import Lead from './Lead.vue';
import data from '../../test/test-lead.json';

const localVue = createLocalVue();
localVue.use(BootstrapVue);
localVue.use(IconsPlugin);

describe('Lead', () => {
    it('matches the snapshot', () => {
        const wrapper = shallowMount(Lead, { localVue, propsData: data });
        expect(wrapper).toMatchSnapshot();
    });
});