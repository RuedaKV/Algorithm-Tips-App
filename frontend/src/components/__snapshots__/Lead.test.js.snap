// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lead matches the snapshot 1`] = `
<div class="lead-box" document_relevance="0.871637" domain="CMS.GOV">
  <b-card-stub bodytag="div" titletag="h4" subtitletag="h6" subtitletextvariant="muted" headertag="div" footertag="div" tag="div" nobody="true">
    <b-card-body-stub bodytag="div" titletag="h4" subtitletag="h6" subtitletextvariant="muted">
      <p class="quote">Quality rating system that gives each nursing home a rating between 1 and 5 stars based on health inspections, staffing levels, and quality measures.</p>
      <div class="source">
        <external-link-stub href="https://www.cms.gov/medicare/provider-enrollment-and-certification/certificationandcomplianc/fsqrs.html">https://www.cms.gov/medicare/provider-enrollment-and-certification/certificationandcomplianc/fsqrs.html</external-link-stub>&nbsp;
        (<external-link-stub href="https://algorithm-tips.s3.us-east-2.amazonaws.com/documents/1420.html" class="cache-link">Cache</external-link-stub>)
      </div>
      <div class="found-by">Found via a search for “<external-link-stub href="https://www.google.com/search?hl=en&amp;q=%22rating%20system%22+site%3A.gov+-site%3A.nih.gov&amp;as_qdr=w1&amp;lr=en&amp;num=100">rating system</external-link-stub>” on June 21st, 2019</div>
      <h5>Additional Info</h5>
      <dl class="row">
        <dt class="col-sm-4">Jurisdiction</dt>
        <dd class="col-sm-8">Federal Agency - Executive</dd>
        <dt class="col-sm-4">Agency</dt>
        <dd class="col-sm-8">Department of Health and Human Services</dd>
        <dt class="col-sm-4">Main Topics</dt>
        <dd class="col-sm-8">Health</dd>
        <dt class="col-sm-4">People &amp; Organizations</dt>
        <dd class="col-sm-8"></dd>
      </dl>
      <h5>Crowd Ratings</h5>
    </b-card-body-stub>
    <b-list-group-stub tag="div" flush="true">
      <b-list-group-item-stub tag="div" target="_self" event="click" routertag="a">
        <div role="button" tabindex="0" class="collapsed" aria-expanded="false" aria-controls="1420Potential for Controversy">
          <b-icon-chevron-right-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-closed"></b-icon-chevron-right-stub>
          <b-icon-chevron-down-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-open"></b-icon-chevron-down-stub> <span class="rating-title">Potential for Controversy</span>
          <b-progress-stub variant="info" precision="0" max="5" value="0" class="w-25 float-right inline-bar">
            <b-progress-bar-stub value="2.6"></b-progress-bar-stub>
          </b-progress-stub> <span class="score-display float-right">2.6 / 5</span>
        </div>
        <b-collapse-stub id="1420Potential for Controversy" tag="div">
          <b-table-lite-stub items="[object Object],[object Object],[object Object],[object Object],[object Object]" value="" striped="true" class="comment-table"></b-table-lite-stub>
        </b-collapse-stub>
      </b-list-group-item-stub>
      <b-list-group-item-stub tag="div" target="_self" event="click" routertag="a">
        <div role="button" tabindex="0" class="collapsed" aria-expanded="false" aria-controls="1420Size of Impact">
          <b-icon-chevron-right-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-closed"></b-icon-chevron-right-stub>
          <b-icon-chevron-down-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-open"></b-icon-chevron-down-stub> <span class="rating-title">Size of Impact</span>
          <b-progress-stub variant="info" precision="0" max="5" value="0" class="w-25 float-right inline-bar">
            <b-progress-bar-stub value="4.6"></b-progress-bar-stub>
          </b-progress-stub> <span class="score-display float-right">4.6 / 5</span>
        </div>
        <b-collapse-stub id="1420Size of Impact" tag="div">
          <b-table-lite-stub items="[object Object],[object Object],[object Object],[object Object],[object Object]" value="" striped="true" class="comment-table"></b-table-lite-stub>
        </b-collapse-stub>
      </b-list-group-item-stub>
      <b-list-group-item-stub tag="div" target="_self" event="click" routertag="a">
        <div role="button" tabindex="0" class="collapsed" aria-expanded="false" aria-controls="1420Negative Societal Impact">
          <b-icon-chevron-right-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-closed"></b-icon-chevron-right-stub>
          <b-icon-chevron-down-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-open"></b-icon-chevron-down-stub> <span class="rating-title">Negative Societal Impact</span>
          <b-progress-stub variant="info" precision="0" max="5" value="0" class="w-25 float-right inline-bar">
            <b-progress-bar-stub value="2.8"></b-progress-bar-stub>
          </b-progress-stub> <span class="score-display float-right">2.8 / 5</span>
        </div>
        <b-collapse-stub id="1420Negative Societal Impact" tag="div">
          <b-table-lite-stub items="[object Object],[object Object],[object Object],[object Object],[object Object]" value="" striped="true" class="comment-table"></b-table-lite-stub>
        </b-collapse-stub>
      </b-list-group-item-stub>
      <b-list-group-item-stub tag="div" target="_self" event="click" routertag="a">
        <div role="button" tabindex="0" class="collapsed" aria-expanded="false" aria-controls="1420Surprising">
          <b-icon-chevron-right-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-closed"></b-icon-chevron-right-stub>
          <b-icon-chevron-down-stub fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" class="when-open"></b-icon-chevron-down-stub> <span class="rating-title">Surprising</span>
          <b-progress-stub variant="info" precision="0" max="5" value="0" class="w-25 float-right inline-bar">
            <b-progress-bar-stub value="1.8"></b-progress-bar-stub>
          </b-progress-stub> <span class="score-display float-right">1.8 / 5</span>
        </div>
        <b-collapse-stub id="1420Surprising" tag="div">
          <b-table-lite-stub items="[object Object],[object Object],[object Object],[object Object],[object Object]" value="" striped="true" class="comment-table"></b-table-lite-stub>
        </b-collapse-stub>
      </b-list-group-item-stub>
    </b-list-group-stub>
  </b-card-stub>
</div>
`;
